pkg/util/gvk-gvr-mapper.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GvkGvrMapper is a thread-safe mapping between GVKs and GVRs.
+type GvkGvrMapper interface {
+	// Add adds a mapping between the given GVK and GVR.
+	Add(gvk schema.GroupVersionKind, gvr schema.GroupVersionResource)
```

Comment:
What if, just before this call, the given gvk was associated with a different gvr? And vice-versa?

Reply:
(A) The reader should be told what happens.

(B) Are you sure that you have made the right choice for what happens?

Reply:
(A) Added

(B) After adding the statement you suggested in the above comment, this scenario "is not supported", no?

Reply:
It is not clear what "This overwrites the GVK <-> GVR mapping if pre-existing" means. Since there is nothing to a mapping but a mapping, what would overwriting a mapping with itself even be? A no-op?

The question is, what happens if this method is invoked at a time when the given gvk is already associated with a _different_ gvr. Also, what if the given gvr is already associated with a _different_ gkv?

Reply:
I just added a comment to #1703, including my expectation that the planned revision will not completely remove the need for mapping between "kind" and "resource". I expect the mapper abstraction to remain. The mapper abstraction's definition should say something about the issue in the original comment of this conversation. It might say "the mapper panics if ..." or it might say that any previous association with either the gvk or gvr is removed or it might be redefined to return an error in the case of partial overlap.